src/main/java/com/pinterest/secor/io/impl/MessagePackSequenceFileReaderWriterFactory.java

```diff
@@ -0,0 +1,165 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pinterest.secor.io.impl;
+
+import com.pinterest.secor.common.LogFilePath;
+import com.pinterest.secor.io.FileReader;
+import com.pinterest.secor.io.FileReaderWriterFactory;
+import com.pinterest.secor.io.FileWriter;
+import com.pinterest.secor.io.KeyValue;
+import com.pinterest.secor.util.FileUtil;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.io.SequenceFile;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePacker;
+import org.msgpack.core.MessageUnpacker;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Sequence file reader writer implementation
+ *
+ * @author Praveen Murugesan (praveen@uber.com)
+ */
+public class MessagePackSequenceFileReaderWriterFactory implements FileReaderWriterFactory {
+    private static final int KAFKA_MESSAGE_OFFSET = 1;
+    private static final int KAFKA_HASH_KEY = 2;
+    private static final byte[] EMPTY_BYTES = new byte[0];
+
+    @Override
+    public FileReader BuildFileReader(LogFilePath logFilePath, CompressionCodec codec) throws Exception {
+        return new MessagePackSequenceFileReader(logFilePath);
+    }
+
+    @Override
+    public FileWriter BuildFileWriter(LogFilePath logFilePath, CompressionCodec codec) throws IOException {
+        return new MessagePackSequenceFileWriter(logFilePath, codec);
+    }
+
+    protected class MessagePackSequenceFileReader implements FileReader {
+        private final SequenceFile.Reader mReader;
+        private final BytesWritable mKey;
+        private final BytesWritable mValue;
+
+        public MessagePackSequenceFileReader(LogFilePath path) throws Exception {
+            Configuration config = new Configuration();
+            Path fsPath = new Path(path.getLogFilePath());
+            FileSystem fs = FileUtil.getFileSystem(path.getLogFilePath());
+            this.mReader = new SequenceFile.Reader(fs, fsPath, config);
+            this.mKey = (BytesWritable) mReader.getKeyClass().newInstance();
+            this.mValue = (BytesWritable) mReader.getValueClass().newInstance();
+        }
+
+        @Override
+        public KeyValue next() throws IOException {
+            if (mReader.next(mKey, mValue)) {
+                MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(mKey.getBytes());
```

Comment:
Is 'unpacker' object expensive to create?  Should the object creation be done in the constructor?